### PR TITLE
[Karpenter Add-on][BREAKING CHANGE] Update for v0.34.0

### PR DESCRIPTION
**Pull request by youngjeong46**
*Issue #, if available:* #924

*Description of changes:*
This PR changes the Karpenter add-on to support up to v0.34.0 (v0.35.0 just released as of the day of the PR, so it will be following as another PR). This is considered a breaking change:
- Refactored properties for the add-on to better reflect the parameters needed for the NodePool and the EC2NodeClass. 
- Ability to deploy the CRDs based on the versions: v0.32.x and up will enforce v1beta1 CRDs (NodePool and EC2NodeClass), while below versions will enforce v1alpha5 CRDs (Provisioner and AWSNodeTemplate)
- Both parameters are optional to be used for either the alpha or the beta CRDS. However, there are error checks implemented in case the wrong one is used (i.e. using `consolidation` with v1beta1 CRDs or `disruption` with v1alpha5 CRDs)
- Addon will no longer support any versions below v0.21.x, as provided in the compatibility matrix from the official website [here](https://karpenter.sh/docs/upgrading/compatibility/). Version Map is used to issue a warning if incompatible version is used for the Kubernetes version, but will proceed with deployment.
- Additional unit tests
- Doc revisions

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by elamaran11** [comment]
@youngjeong46 Nice PR, lof of work. Does this also fix #859 and #893 Issues?

**Comment by youngjeong46** [comment]
> @youngjeong46 Nice PR, lof of work. Does this also fix #859 and #893 Issues?

No, not yet. I'm trying to figure out the exact ask (at least for 893). Given this PR, it would most likely be a fast follow (along with 0.35 update that came out last night).

File: docs/addons/karpenter.md [hunk 1]
```diff
@@ -16,66 +16,63 @@ Karpenter works by:
...
   .addOns(karpenterAddOn)
   .build(app, 'my-stack-name');
 ```
 
-The add-on automatically sets the following Helm Chart [values](https://github.com/aws/karpenter/tree/main/charts/karpenter#values), and it is **highly recommended** not to pass these values in (as it will result in errors):
-- settings.aws.defaultInstanceProfile
-- settings.aws.clusterEndpoint
-- settings.aws.clusterName
-- settings.aws.interruptionQueueName (if interruption handling is enabled)
+The add-on automatically sets the following Helm Chart [values](https://github.com/aws/karpenter-provider-aws/tree/main/charts/karpenter#values), and it is **highly recommended** not to pass these values in (as it will result in errors):
```
Comment (shapirov103) [inline docs/addons/karpenter.md:73]: why would passing the values result in errors? if the values are correct, it should succeed.

File: docs/addons/karpenter.md [hunk 2]
```diff
@@ -94,57 +91,88 @@ blueprints-addon-karpenter-54fd978b89-hclmp   2/2     Running   0          99m
...
+4. Deploys Karpenter helm chart in the `karpenter` namespace, configuring the cluster name, endpoint, Instance Profile, and others necessary for functional addon.
+5. If the user provides NodePoolSpec (and EC2NodeClassSpec), the addon will provisions a default Karpenter NodePool and EC2NodeClass CRDs. NodePoolSpec requires [requirements](https://karpenter.sh/docs/concepts/nodepools/#spectemplatespecrequirements) while EC2NodeClassSpec requires subnets and security groups. Based on what version of Karpenter you provide, you will need either `subnetSelector` and `securityGroupSelector` (for versions v0.31.x or down), or `subnetSelectorTerms` and `securityGroupSelectorTerms` (for versions v0.32.x and up).
```
Comment (shapirov103) [inline docs/addons/karpenter.md:95]: these attributes should be lower-case similar to the rest of the attrbutes based on the naming conventions: attributes/functions lower case, types and class name upper case.

File: docs/addons/karpenter.md [hunk 3]
```diff
@@ -94,57 +91,88 @@ blueprints-addon-karpenter-54fd978b89-hclmp   2/2     Running   0          99m
...
-5. Interruption Handling, which is a native way to handle interruption due to involuntary interruption events, is supported on versions 0.19.0 and later. For interruption handling in the earlier versions, Karpenter supports using AWS Node Interruption Handler (which you will need to add as an add-on and ***must be in add-on array after the Karpenter add-on*** for it to work.
-6. Karpenter allows overrides of the default "Name" tag but does not allow overrides to restricted domain (such as "karpenter.sh", "karpenter.k8s.aws", and "kubernetes.io/cluster").
+1. EKS Blueprints will only support minimum Karpenter version that matches the supporter EKS Kubernetes version. Please see the compatibility matrix [here](https://karpenter.sh/docs/upgrading/compatibility/). If you provide incompatible version (i.e. providing version 0.27.x for EKS version 1.27), you will see warnings in the logs but will proceed deployment. You will run into compatibility issues.
+2. EKS Blueprints npm v1.13.x introduces the following breaking changes:
```
Comment (shapirov103) [inline docs/addons/karpenter.md:100]: this will 1.14 and above.

File: docs/addons/karpenter.md [hunk 4]
```diff
@@ -213,35 +241,6 @@ Karpenter, starting from the OCI registry versions, will untar the files under `
...
-
-3. Starting with v0.17.0, Karpenter's Helm chart package is stored in OCI (Open Container Initiative) registry. With this change, [charts.karpenter.sh](https://charts.karpenter.sh/) is no longer updated to preserve older versions. You have to adjust for the following:
-
-* The full URL needs to be present (including 'oci://').
-* You need to append a `v` to the version number (i.e. v0.17.0, not 0.17.0)
+1. Certain upgrades require reapplying the CRDs since Helm does not maintain the lifecycle of CRDs. Please see the [official documentations](https://karpenter.sh/docs/upgrading/upgrade-guide/) for details.
 
-4. Starting with v0.22.0, Karpenter will no longer work on Kubernetes version prior to 1.21. Either upgrade your Kubernetes to 1.21 or later version and apply Karpenter, or use prior Karpenter versions.
+2. Starting v0.32.0, Karpenter introduces the new beta APIs (v1beta1). While you can use both v1beta1 and v1alpha5 CRDs for v0.32, starting in v0.33, v1alpha5 CRDs are obsolete. If you want to upgrade to the latest, ensure that you have upgraded to v0.32.x first before proceeding.
```
Comment (shapirov103) [inline docs/addons/karpenter.md:246]: trying to figure out exact upgrade steps for the current blueprint customers. Does this paragraph implies that current customer should:
1. upgrade to 0.32.x 
2. upgrade to the latest

Are the any other steps? My assumption is that CRDs will be upgrade (new set of CRDs based on the versoin) and the above approach should work.

File: docs/addons/karpenter.md [hunk 5]
```diff
@@ -213,35 +241,6 @@ Karpenter, starting from the OCI registry versions, will untar the files under `
...
-
-The property is changed to align with the naming convention of the provisioner, and to allow multiple operators (In vs NotIn). The values correspond similarly between the two, with type change being the only difference.
-
-2. Certain upgrades require reapplying the CRDs since Helm does not maintain the lifecycle of CRDs. Please see the [official documentations](hhttps://karpenter.sh/docs/upgrading/upgrade-guide/) for details.
-
-3. Starting with v0.17.0, Karpenter's Helm chart package is stored in OCI (Open Container Initiative) registry. With this change, [charts.karpenter.sh](https://charts.karpenter.sh/) is no longer updated to preserve older versions. You have to adjust for the following:
-
-* The full URL needs to be present (including 'oci://').
-* You need to append a `v` to the version number (i.e. v0.17.0, not 0.17.0)
+1. Certain upgrades require reapplying the CRDs since Helm does not maintain the lifecycle of CRDs. Please see the [official documentations](https://karpenter.sh/docs/upgrading/upgrade-guide/) for details.
```
Comment (shapirov103) [inline docs/addons/karpenter.md:244]: is this something that blueprints can handle seemlessly. Please see my comment on the next paragraph as well.

File: lib/addons/karpenter/index.ts [hunk 6]
```diff
@@ -39,124 +65,203 @@ export interface EbsVolumeMapping {
...
-    securityGroupTags?: Values,
-
-    /**
-     * AMI Family: If provided, Karpenter will automatically query the appropriate EKS optimized AMI via AWS Systems Manager
+     * This is the top level nodepool specification. Nodepools launch nodes in response to pods that are unschedulable. 
+     * A single nodepool is capable of managing a diverse set of nodes. 
+     * Node properties are determined from a combination of nodepool and pod scheduling constraints.
      */
-    amiFamily?: "AL2" | "Bottlerocket" | "Ubuntu" | "Windows2019" | "Windows2022"
+    NodePoolSpec?: {
```
Comment (shapirov103) [inline lib/addons/karpenter/index.ts:72]: attributes lower case - `nodePoolSpec` just like the rest of the attributes. 

You can extract a type like 
`export type NodePoolSpec ...`, then use `nodePoolSpec: NodePoolSpec`.

The above approach with exported type allows configuring objects of type NodePoolSpec in customer projects that import blueprints. 

The way you declared it now - it creates an anonymous type that customers cannot use directly (not exported) unless they create an object of type `KarpenterAddOnProps`. 

Some customers brought it up, that if they want to reuse karpenter across several blueprints and just modify for example nodePoolSpec, it becomes a problem as they need to clone the whole outer structure.

File: lib/addons/karpenter/index.ts [hunk 7]
```diff
@@ -39,124 +65,203 @@ export interface EbsVolumeMapping {
...
+     * It contains configuration necessary to launch instances in AWS. 
      */
-    ttlSecondsAfterEmpty?: number,
-
-    /**
-     * Priority given to the provisioner when the scheduler considers which provisioner
-     * to select. Higher weights indicate higher priority when comparing provisioners.
-     */
-    weight?: number,
+    EC2NodeClassSpec?: {
```
Comment (shapirov103) [inline lib/addons/karpenter/index.ts:175]: `ec2NodeClassSpec`

File: lib/addons/karpenter/index.ts [hunk 8]
```diff
@@ -332,46 +474,121 @@ export class KarpenterAddOn extends HelmAddOn {
...
-        if ((Object.keys(subnetTags).length > 0) && (Object.keys(sgTags).length > 0)){
-            const provisioner = cluster.addManifest('default-provisioner', {
-                apiVersion: 'karpenter.sh/v1alpha5',
-                kind: 'Provisioner',
-                metadata: { name: 'default' },
-                spec: {
-                    providerRef: {
-                        name: "default"
+        // Deploy Provisioner (Alpha) or NodePool (Beta) CRD based on the Karpenter Version
+        if (this.options.NodePoolSpec){
```
Comment (shapirov103) [inline lib/addons/karpenter/index.ts:478]: code style: I observe cases when the template is quite different but injected values are the same. Such code blocks with if/then take up space and make code script style that is tough to navigate. 
Consider using kubectlProvider.addManifest which allows passing replacement values to the templates. example [here](https://github.com/aws-quickstart/cdk-eks-blueprints/blob/main/lib/addons/cloudwatch-adot-addon/index.ts#L100) and similar in the codebase.

In that style you can load two templates and pass the same set of values to get the right manifest. 

Can be deferred.

File: lib/builders/windows-builder.ts [hunk 9]
```diff
@@ -121,27 +121,34 @@ const defaultOptions: WindowsOptions = {
 
 const defaultKarpenterProps: addons.KarpenterAddOnProps = {
```
Comment (shapirov103) [inline lib/builders/windows-builder.ts:122]: let's run `make lint` to remove this warning that defaultKarpenterProps is assigned a value but never used. 
Also is it an indication of a defect? why default props are never used?

**Review by shapirov103 — COMMENTED** [review]
@youngjeong46 tremendous work, a lot of changes and new features. 
Please see my review comments. I will then kick off a test on 1.28. The move the default EKS version to 1.29 in a separate PR and retest before release on 1.29. 
Great work!

File: docs/addons/karpenter.md (hunk 4 above)
Comment (jsamuel1) [reply docs/addons/karpenter.md:246]: If we handle the translation of the CRDs to the new versions, and we have linting to ensure only valid config, what is the advantage/benefit of the middle step of upgrading to 0.32.x?
For customers self-managing Karpenter - this allows a transition, whereby both alpha and beta  CRDs are active, and therefore they may upgrade, then upgrade config to beta, then upgrade again.

File: docs/addons/karpenter.md (hunk 2 above)
Comment (youngjeong46) [reply docs/addons/karpenter.md]: lower cased ec2NodeClass and nodePoolSpec - the others (subnet, etc) are attributes inside those and should remain as so. addressed.

File: lib/addons/karpenter/index.ts (hunk 7 above)
Comment (youngjeong46) [reply lib/addons/karpenter/index.ts]: addressed

File: lib/builders/windows-builder.ts (hunk 9 above)
Comment (youngjeong46) [reply lib/builders/windows-builder.ts:122]: This was leftovers from previous commit from another PR. Can remove as I don't see the necessity.

Comment (youngjeong46) [reply lib/builders/windows-builder.ts]: addressed

File: docs/addons/karpenter.md (hunk 3 above)
Comment (youngjeong46) [reply docs/addons/karpenter.md]: revised. addressed.

File: docs/addons/karpenter.md (hunk 1 above)
Comment (youngjeong46) [reply docs/addons/karpenter.md:73]: I can change the wording to recommend not to pass them in

Comment (youngjeong46) [reply docs/addons/karpenter.md]: done

File: lib/addons/karpenter/index.ts (hunk 6 above)
Comment (youngjeong46) [reply lib/addons/karpenter/index.ts]: exported outside to SPI

File: docs/addons/karpenter.md (hunk 4 above)
Comment (shapirov103) [reply docs/addons/karpenter.md:246]: it seems easier to just create a separate Addon for Karpenter post 0.32 and just focus on beta and keep the existing addon as is and deprecate it. The upgrade path: remove the existing addon and add Karpenter Beta addon. If that path does not work, then we can discuss.

Comment (shapirov103) [reply docs/addons/karpenter.md:246]: Given the amount of rework, it does not appear to be the optimal path for the current release. Refactoring can be deferred. Ideally, this type of decisions are best to discuss ahead of the implementation. If we release in the current state, it is not a one-way door, we can move forward.

File: docs/addons/karpenter.md (hunk 5 above)
Comment (youngjeong46) [reply docs/addons/karpenter.md:244]: This should be taken care of. I can remove this wording.

File: lib/addons/karpenter/index.ts (hunk 8 above)
Comment (youngjeong46) [reply lib/addons/karpenter/index.ts]: will defer.

File: docs/addons/karpenter.md (hunk 4 above)
Comment (youngjeong46) [reply docs/addons/karpenter.md:246]: I think we can defer this.

File: lib/spi/types.ts [hunk 10]
```diff
@@ -24,15 +26,251 @@ export type Values = {
     [key: string]: any;
 };
 
+/**
+ * Utility type for Karpenter requirement values
+ */
+export type reqValues = {
```
Comment (shapirov103) [inline lib/spi/types.ts:32]: minor: Type names start with uppercase - ReqValues. What does Req mean here? required?

File: lib/spi/types.ts [hunk 11]
```diff
@@ -24,15 +26,251 @@ export type Values = {
...
+    key: string,
```
Comment (shapirov103) [inline lib/spi/types.ts:33]: is there are reason why Karpenter specific type is plance in spi as opposed to a regular export from index.ts?

**Review by shapirov103 — COMMENTED** [review]
See my last comment, I will kick off e2e once addressed since code changes invalidate the e2e approval.

File: lib/spi/types.ts (hunk 11 above)
Comment (youngjeong46) [reply lib/spi/types.ts]: moved the types only relevant to Karpenter back to `index.ts`

**Comment by shapirov103** [comment]
@youngjeong46 getting failure on jest test:


FAIL test/karpenter.test.ts
--
214 | ● Test suite failed to run
215 |  
216 | test/karpenter.test.ts:7:10 - error TS2724: '"../lib/spi"' has no exported member named 'reqValues'. Did you mean 'Values'?
217 |  
218 | 7 import { reqValues } from "../lib/spi";

**Review by shapirov103 — APPROVED** [review]
LGTM

**Pull request merged**